File: paimon-core/src/main/java/org/apache/paimon/partition/PartitionTimeExtractor.java
/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.apache.paimon.partition;

import org.apache.paimon.data.Timestamp;

import javax.annotation.Nullable;

import java.time.LocalDate;
import java.time.LocalDateTime;
import java.time.LocalTime;
import java.time.format.DateTimeFormatter;
import java.time.format.DateTimeFormatterBuilder;
import java.time.format.DateTimeParseException;
import java.time.format.ResolverStyle;
import java.time.format.SignStyle;
import java.time.temporal.ChronoField;
import java.util.ArrayList;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Locale;
import java.util.Objects;

import static java.time.temporal.ChronoField.DAY_OF_MONTH;
import static java.time.temporal.ChronoField.HOUR_OF_DAY;
import static java.time.temporal.ChronoField.MINUTE_OF_HOUR;
import static java.time.temporal.ChronoField.MONTH_OF_YEAR;
import static java.time.temporal.ChronoField.SECOND_OF_MINUTE;
import static java.time.temporal.ChronoField.YEAR;

/** Time extractor to extract time from partition values. */
public class PartitionTimeExtractor {

    private static final DateTimeFormatter TIMESTAMP_FORMATTER =
            new DateTimeFormatterBuilder()
                    .appendValue(YEAR, 1, 10, SignStyle.NORMAL)
                    .appendLiteral('-')
                    .appendValue(MONTH_OF_YEAR, 1, 2, SignStyle.NORMAL)
                    .appendLiteral('-')
                    .appendValue(DAY_OF_MONTH, 1, 2, SignStyle.NORMAL)
                    .optionalStart()
                    .appendLiteral(" ")
                    .appendValue(HOUR_OF_DAY, 1, 2, SignStyle.NORMAL)
                    .appendLiteral(':')
                    .appendValue(MINUTE_OF_HOUR, 1, 2, SignStyle.NORMAL)
                    .appendLiteral(':')
                    .appendValue(SECOND_OF_MINUTE, 1, 2, SignStyle.NORMAL)
                    .optionalStart()
                    .appendFraction(ChronoField.NANO_OF_SECOND, 1, 9, true)
                    .optionalEnd()
                    .optionalEnd()
                    .toFormatter()
                    .withResolverStyle(ResolverStyle.LENIENT);

    private static final DateTimeFormatter DATE_FORMATTER =
            new DateTimeFormatterBuilder()
                    .appendValue(YEAR, 1, 10, SignStyle.NORMAL)
                    .appendLiteral('-')
                    .appendValue(MONTH_OF_YEAR, 1, 2, SignStyle.NORMAL)
                    .appendLiteral('-')
                    .appendValue(DAY_OF_MONTH, 1, 2, SignStyle.NORMAL)
                    .toFormatter()
                    .withResolverStyle(ResolverStyle.LENIENT);

    @Nullable private final String pattern;
    @Nullable private final String formatter;

    public PartitionTimeExtractor(@Nullable String pattern, @Nullable String formatter) {
        this.pattern = pattern;
        this.formatter = formatter;
    }

    public LocalDateTime extract(LinkedHashMap<String, String> spec) {
        return extract(new ArrayList<>(spec.keySet()), new ArrayList<>(spec.values()));
    }

    public LocalDateTime extract(List<String> partitionKeys, List<?> partitionValues) {
        // Try to extract directly from typed partition values (DATE/TIMESTAMP)
        if (pattern == null && partitionValues.size() == 1) {
            Object value = partitionValues.get(0);
            LocalDateTime directExtracted = extractFromTypedValue(value);
            if (directExtracted != null) {
                return directExtracted;
            }
        }

        // Handle pattern-based extraction with typed values
        if (pattern != null) {
            for (int i = 0; i < partitionKeys.size(); i++) {
                Object value = partitionValues.get(i);
                LocalDateTime directExtracted = extractFromTypedValue(value);
                if (directExtracted != null) {
                    // For typed values with pattern, replace with formatted string
                    String formattedValue = formatTypedValue(directExtracted);
                    if (formattedValue != null) {
                        String tempPattern =
                                pattern.replaceAll("\\$" + partitionKeys.get(i), formattedValue);
                        // If pattern only contains this one variable, we can use the extracted
                        // value
                        if (!tempPattern.contains("$")) {
                            return toLocalDateTime(tempPattern, this.formatter);
                        }
                    }
                }
            }
        }

        // Fall back to string-based extraction
        String timestampString;
        if (pattern == null) {
            timestampString = partitionValues.get(0).toString();
        } else {
            timestampString = pattern;
            for (int i = 0; i < partitionKeys.size(); i++) {
                timestampString =
                        timestampString.replaceAll(
                                "\\$" + partitionKeys.get(i), partitionValues.get(i).toString());
            }
        }
        return toLocalDateTime(timestampString, this.formatter);
    }

    private static LocalDateTime toLocalDateTime(
            String timestampString, @Nullable String formatterPattern) {

        if (formatterPattern == null) {
            return PartitionTimeExtractor.toLocalDateTimeDefault(timestampString);
        }
        DateTimeFormatter dateTimeFormatter =
                DateTimeFormatter.ofPattern(Objects.requireNonNull(formatterPattern), Locale.ROOT);
        try {
            return LocalDateTime.parse(timestampString, Objects.requireNonNull(dateTimeFormatter));
        } catch (DateTimeParseException e) {
            return LocalDateTime.of(
                    LocalDate.parse(timestampString, Objects.requireNonNull(dateTimeFormatter)),
                    LocalTime.MIDNIGHT);
        }
    }

    public static LocalDateTime toLocalDateTimeDefault(String timestampString) {
        try {
            return LocalDateTime.parse(timestampString, TIMESTAMP_FORMATTER);
        } catch (DateTimeParseException e) {
            return LocalDateTime.of(
                    LocalDate.parse(timestampString, DATE_FORMATTER), LocalTime.MIDNIGHT);
        }
    }

    /**
     * Extract LocalDateTime from typed partition values (Integer for DATE, Timestamp for
     * TIMESTAMP).
     *
     * @param value the partition value object
     * @return LocalDateTime if the value is a supported type, null otherwise
     */
    @Nullable
    private static LocalDateTime extractFromTypedValue(Object value) {
        if (value == null) {
            return null;
        }

        // Handle INTEGER type - represents days since epoch (DATE type)
        if (value instanceof Integer) {
            int daysSinceEpoch = (Integer) value;
            return LocalDate.ofEpochDay(daysSinceEpoch).atStartOfDay();
        }

        // Handle Paimon Timestamp type
        if (value instanceof Timestamp) {
            Timestamp timestamp = (Timestamp) value;
            return timestamp.toLocalDateTime();
        }

        // Handle java.sql.Date
        if (value instanceof java.sql.Date) {
            return ((java.sql.Date) value).toLocalDate().atStartOfDay();
        }

        // Handle java.sql.Timestamp
        if (value instanceof java.sql.Timestamp) {
            return ((java.sql.Timestamp) value).toLocalDateTime();
        }

        // Handle java.time.LocalDate
        if (value instanceof LocalDate) {
            return ((LocalDate) value).atStartOfDay();
        }

        // Handle java.time.LocalDateTime
        if (value instanceof LocalDateTime) {
            return (LocalDateTime) value;
        }

        return null;
    }

    /**
     * Format a LocalDateTime to string for pattern replacement.
     *
     * @param dateTime the LocalDateTime to format
     * @return formatted string, or null if input is null
     */
    @Nullable
    private static String formatTypedValue(LocalDateTime dateTime) {
        if (dateTime == null) {
            return null;
        }
        // Use ISO format: yyyy-MM-dd HH:mm:ss or yyyy-MM-dd if time is midnight
        if (dateTime.toLocalTime().equals(LocalTime.MIDNIGHT)) {
            return dateTime.toLocalDate().toString();
        }
        return dateTime.toString().replace("T", " ");
    }
}

File: paimon-python/pypaimon/partition/partition_time_extractor.py
# Licensed to the Apache Software Foundation (ASF) under one
# or more contributor license agreements.  See the NOTICE file
# distributed with this work for additional information
# regarding copyright ownership.  The ASF licenses this file
# to you under the Apache License, Version 2.0 (the
# "License"); you may not use this file except in compliance
# with the License.  You may obtain a copy of the License at
#
#   http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing,
# software distributed under the License is distributed on an
# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
# KIND, either express or implied.  See the License for the
# specific language governing permissions and limitations
# under the License.

"""
Time extractor to extract datetime from partition values.

This mirrors the Java implementation at:
  org.apache.paimon.partition.PartitionTimeExtractor
"""

import re
from datetime import datetime, time as dt_time
from typing import Dict, List, Optional


# Default timestamp formats (lenient parsing, mirrors Java TIMESTAMP_FORMATTER)
_DEFAULT_TIMESTAMP_FORMATS = [
    "%Y-%m-%d %H:%M:%S.%f",
    "%Y-%m-%d %H:%M:%S",
    "%Y-%m-%dT%H:%M:%S.%f",
    "%Y-%m-%dT%H:%M:%S",
    "%Y-%m-%d %H:%M",
    "%Y-%m-%d",
]

# Java DateTimeFormatter → Python strptime token mapping.
# Note: Python's %f is microseconds (6 digits); Java's SSS is milliseconds (3 digits).
# This works for parsing because strptime(%f) accepts 1-6 digits, but the semantic
# precision differs. Callers should be aware of this when formatting output.
_JAVA_TO_PYTHON_TOKENS = {
    "yyyy": "%Y",
    "yy": "%y",
    "MM": "%m",
    "dd": "%d",
    "HH": "%H",
    "mm": "%M",
    "SSS": "%f",
    "SS": "%f",
    "ss": "%S",
}

# Regex that matches Java tokens (longest first) or quoted literals
_JAVA_TOKEN_RE = re.compile(
    r"'([^']*)'|(" + "|".join(re.escape(k) for k in sorted(_JAVA_TO_PYTHON_TOKENS, key=len, reverse=True)) + r")"
)


def _java_to_python_format(java_fmt: str) -> str:
    """Convert Java DateTimeFormatter pattern to Python strptime format.

    Uses regex tokenization to avoid overlapping-replacement issues with
    sequential str.replace().
    """
    def _replace_token(m):
        if m.group(1) is not None:
            return m.group(1)
        return _JAVA_TO_PYTHON_TOKENS[m.group(2)]

    # Replace known tokens; leave unmatched characters (literals like '-', ' ', 'T') as-is
    result = _JAVA_TOKEN_RE.sub(_replace_token, java_fmt)
    return result


class PartitionTimeExtractor:
    """
    Extracts a datetime from partition key values.

    Supports two modes:
    1. No pattern: uses the first partition value directly as a timestamp string.
    2. With pattern: replaces $<key> placeholders with partition values, then parses.

    The optional ``formatter`` is a strptime-compatible format string used to parse
    the resulting timestamp string.
    """

    def __init__(self, pattern: Optional[str] = None, formatter: Optional[str] = None):
        self._pattern = pattern
        self._formatter = formatter

    def extract(self, partition_keys: List[str], partition_values: List[object]) -> datetime:
        """
        Extract a datetime from the given partition keys and values.

        Args:
            partition_keys: Ordered list of partition column names.
            partition_values: Ordered list of partition column values (stringifiable).

        Returns:
            The extracted datetime.

        Raises:
            ValueError: If the timestamp string cannot be parsed.
        """
        timestamp_string = self._build_timestamp_string(partition_keys, partition_values)
        return self._to_local_datetime(timestamp_string)

    def extract_from_spec(self, spec: Dict[str, str]) -> datetime:
        """
        Extract a datetime from a partition spec dictionary.

        Args:
            spec: Partition specification {key: value}.

        Returns:
            The extracted datetime.
        """
        keys = list(spec.keys())
        values = list(spec.values())
        return self.extract(keys, values)

    def _build_timestamp_string(self, partition_keys: List[str], partition_values: List[object]) -> str:
        if self._pattern is None:
            return str(partition_values[0])
        result = self._pattern
        for i, key in enumerate(partition_keys):
            result = re.sub(r'\$' + re.escape(key), str(partition_values[i]), result)
        return result

    def _to_local_datetime(self, timestamp_string: str) -> datetime:
        if self._formatter is not None:
            py_format = _java_to_python_format(self._formatter)
            return self._parse_with_formatter(timestamp_string, py_format)
        return self._parse_default(timestamp_string)

    @staticmethod
    def _parse_with_formatter(timestamp_string: str, formatter: str) -> datetime:
        """Parse using the converted Python strptime pattern.

        Mirrors Java behavior: tries LocalDateTime.parse first, then falls back
        to LocalDate.parse (date-only) with the same formatter pattern.
        """
        try:
            return datetime.strptime(timestamp_string, formatter)
        except ValueError:
            # Fallback: strip time directives (%H, %M, %S, %f) and surrounding
            # literals to get a date-only format, mirroring Java's LocalDate.parse fallback.
            date_only_format = re.split(r'(?=%H|%M|%S|%f)', formatter)[0].rstrip()
            # Also strip any trailing non-directive separators (e.g. trailing space or 'T')
            date_only_format = date_only_format.rstrip(' T-:')
            if date_only_format and date_only_format != formatter:
                parsed_date = datetime.strptime(timestamp_string, date_only_format).date()
                return datetime.combine(parsed_date, dt_time.min)
            raise

    @staticmethod
    def _parse_default(timestamp_string: str) -> datetime:
        """Parse using default formats (lenient)."""
        for fmt in _DEFAULT_TIMESTAMP_FORMATS:
            try:
                return datetime.strptime(timestamp_string, fmt)
            except ValueError:
                continue
        raise ValueError(
            f"Cannot parse timestamp string '{timestamp_string}' with any default format. "
            f"Tried: {_DEFAULT_TIMESTAMP_FORMATS}"
        )
